components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java

```diff
@@ -702,4 +599,137 @@ public boolean isAuthorizedClient(OAuthTokenReqMessageContext tokReqMsgCtx) thro
+            if (getAccessTokenExpiryTimeMillis(existingToken) == 0) {
+                // Token is expired. Clear it from cache.
+                removeFromCache(cacheKey, consumerKey, existingToken);
+            }
+        }
+        return existingToken;
+    }
+
+    private void removeFromCache(OAuthCacheKey cacheKey, String consumerKey, AccessTokenDO existingAccessTokenDO) {
+        oauthCache.clearCacheEntry(cacheKey);
+        if (log.isDebugEnabled()) {
+            if (IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.ACCESS_TOKEN)) {
+                log.debug("Access token(hashed) " + DigestUtils.sha256Hex(existingAccessTokenDO
+                        .getAccessToken()) + " is expired. Therefore cleared it from cache and marked" +
+                        " it as expired in database");
+            } else {
+                log.debug("Existing access token for client: " + consumerKey + " is expired. " +
+                        "Therefore cleared it from cache and marked it as expired in database");
+            }
+        }
+    }
+
+    private boolean isRefreshTokenValid(AccessTokenDO existingAccessTokenDO, long validityPeriod, String consumerKey) {
+        if (existingAccessTokenDO != null) {
+            long refreshTokenExpireTime = OAuth2Util.getRefreshTokenExpireTimeMillis(existingAccessTokenDO);
+            if (OAuthConstants.TokenStates.TOKEN_STATE_ACTIVE.equals(existingAccessTokenDO.getTokenState())) {
+                if (!isRefreshTokenExpired(validityPeriod, refreshTokenExpireTime)) {
+                    if (log.isDebugEnabled()) {
+                        if (IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.ACCESS_TOKEN)) {
+                            log.debug("Existing access token: " + existingAccessTokenDO.getAccessToken() +
+                                    " has expired, but refresh token:" +
+                                    existingAccessTokenDO.getRefreshToken() + " is still valid for client: " +
+                                    consumerKey + ". Remaining time: " + refreshTokenExpireTime +
+                                    "ms. Using existing refresh token.");
+
+                        } else {
+                            log.debug("Existing access token has expired, but refresh token is still valid " +
+                                    "for client: " + consumerKey + ". Remaining time: " +
+                                    refreshTokenExpireTime + "ms. Using existing refresh token.");
+                        }
+                    }
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private boolean isRefreshTokenExpired(long validityPeriod, long refreshTokenExpireTime) {
+        if (refreshTokenExpireTime < 0) {
+            // refresh token has infinite validity
+            return false;
+        }
+        if (refreshTokenExpireTime > 0 && refreshTokenExpireTime > validityPeriod) {
```

Comment:
Can be simplified as

 return !(refreshTokenExpireTime > 0 && refreshTokenExpireTime > validityPeriod);